### PR TITLE
Parse metrics tests

### PR DESCRIPTION
**Pull request by Darkxzie**
## Description
Unit Tests for  parse.py and parse_metrics.py covering edge cases, end to end functionality and general coverage.

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepretro/utils/parse.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        }
+
+    def _parse_node(
+        self,
+        data: RouteNode,
+        steps: list[Step],
+        dependencies: MutableMapping[str, list[str]],
+        parent_id: Optional[int],
+    ) -> None:
+        """Parse one route node and recurse into its precursor children."""
```
Comment (ARY2260) [inline deepretro/utils/parse.py]: please add proper doctrings for helper methods as well

File: deepretro/utils/parse.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+ParseOutput = dict[str, Any]
+
+
+def _default_basic_molecules() -> Collection[str]:
+    """Return the configured set of commercially basic molecules."""
+    from deepretro.utils.variables import BASIC_MOLECULES
+
+    return BASIC_MOLECULES
+
+
```
Comment (ARY2260) [inline deepretro/utils/parse.py]: please remove extra line

File: deepretro/utils/parse_metrics.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+            active_logger = logger
+        if active_logger is not None:
+            active_logger.error(
+                "Error in metric calculation",
+                function=function_name,
+                error=str(exc),
+            )
+
+
+def get_reaction_type(mol1: str, mol2: str, model_path: str) -> tuple[str, int]:
```
Comment (ARY2260) [inline deepretro/utils/parse_metrics.py:221]: add proper doctrings

**Review by ARY2260 — CHANGES_REQUESTED** [review]
Once minor changes are done, looks good to me

File: deepretro/utils/parse_metrics.py (hunk 3 above)
Comment (Darkxzie) [reply deepretro/utils/parse_metrics.py:221]: added.

File: deepretro/utils/parse.py (hunk 2 above)
Comment (Darkxzie) [reply deepretro/utils/parse.py]: done.

File: deepretro/utils/parse.py (hunk 1 above)
Comment (Darkxzie) [reply deepretro/utils/parse.py]: done.

**Comment by Darkxzie** [comment]
> Once minor changes are done, looks good to me

Minor changes have been made.

File: deepretro/utils/parse.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+from deepretro.utils.parse_metrics import calc_scalability_index
+from deepretro.utils.utils_molecule import calc_chemical_formula, calc_mol_wt
+
+RouteNode = Mapping[str, Any]
+Step = dict[str, Any]
+DependencyMap = dict[str, list[str]]
+ParseOutput = dict[str, Any]
+
+
+def _default_basic_molecules() -> Collection[str]:
```
Comment (riya-singh28) [inline deepretro/utils/parse.py]: @Darkxzie This function is redundant since it only imports and returns BASIC_MOLECULES without performing any modifications or processing.

File: deepretro/utils/parse.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+            Existing dependency accumulator used by recursive callers.
+        parent_id : int, optional
+            One-based step id of the parent reaction.
+
+        Returns
+        -------
+        dict[str, Any]
+            Dictionary with ``dependencies`` and ``steps`` keys.
+        """
+        del include_metadata
```
Comment (riya-singh28) [inline deepretro/utils/parse.py]: @Darkxzie, I don’t think removing `include_metadata` is a good idea. If a globally scoped object is passed as input, deleting `include_metadata` here could unintentionally affect that object as well. It would be safer to avoid removing it.

File: deepretro/utils/parse.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        self.mass_calculator = (
+            calc_mol_wt if mass_calculator is None else mass_calculator
+        )
+        self.scalability_calculator = (
+            calc_scalability_index
+            if scalability_calculator is None
+            else scalability_calculator
+        )
+
+    def parse_step(
```
Comment (riya-singh28) [inline deepretro/utils/parse.py:83]: Add a small example here.

File: deepretro/utils/parse.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        parent_step["reactionmetrics"][0]["scalabilityindex"] = (
+            self.scalability_calculator(smiles, parent_step["products"][0]["smiles"])
+        )
+
+    def _molecule_entry(self, smiles: str, metadata_key: str) -> dict[str, Any]:
+        """Build a molecule entry with the metadata key required by the schema."""
+        return {
+            "smiles": smiles,
+            metadata_key: {
+                "name": "",
```
Comment (riya-singh28) [inline deepretro/utils/parse.py:310]: Why is this name field empty?

File: deepretro/utils/parse.py [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+        smiles = self._smiles(data)
+        parent_step = steps[parent_id - 1]
+        molecule_type = "reagents" if smiles in self.basic_molecules else "reactants"
+        metadata_key = f"{molecule_type[:-1]}_metadata"
+        parent_step[molecule_type].append(self._molecule_entry(smiles, metadata_key))
+        parent_step["reactionmetrics"][0]["scalabilityindex"] = (
+            self.scalability_calculator(smiles, parent_step["products"][0]["smiles"])
+        )
+
+    def _molecule_entry(self, smiles: str, metadata_key: str) -> dict[str, Any]:
```
Comment (riya-singh28) [inline deepretro/utils/parse.py:288]: Add proper docstrings

File: deepretro/utils/parse.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            "conditions": [],
+            "reactionmetrics": [
+                {
+                    "scalabilityindex": "",
+                    "closestliterature": "",
+                }
+            ],
+        }
+
+    def _attach_to_parent(
```
Comment (riya-singh28) [inline deepretro/utils/parse.py:249]: Please add proper docstrings

File: deepretro/utils/parse_metrics.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+        mol2 : str
+            Product SMILES.
+
+        Returns
+        -------
+        tuple[str, int]
+            Reaction label and classifier encoding index. Unknown or failed
+            predictions return ``("Unknown Reaction", -1)``.
+        """
+        try:
```
Comment (riya-singh28) [inline deepretro/utils/parse_metrics.py:143]: I think this step is redundant here. Not logging FileNotFoundError will make debugging difficult.
`except FileNotFoundError:
            return UNKNOWN_REACTION`

File: docs/source/package/deepretro.utils.parse_metrics.rst [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+Testing and Dependency Injection
+--------------------------------
+
+The calculator accepts injected dependencies so tests do not need to read a
+joblib model or compute RDKit fingerprints:
+
+.. code-block:: python
+
+   class FakeClassifier:
```
Comment (riya-singh28) [inline docs/source/package/deepretro.utils.parse_metrics.rst]: You can rename it to Dummy instead of Fake

File: deepretro/utils/parse_metrics.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+    return utils_molecule.compute_fingerprint(smiles)
+
+
+class ReactionMetricCalculator:
+    """
+    Calculate scalability metrics for parsed route steps.
+
+    Examples
+    --------
+    >>> calculator = ReactionMetricCalculator(model_path="")
```
Comment (riya-singh28) [inline deepretro/utils/parse_metrics.py:54]: Please add an example with model_path and model_loader inputs. You can make an assets folder below tests to include any needed files for example.

File: deepretro/utils/parse_metrics.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        self.model_path = (
+            _rxn_classification_model_path() if model_path is None else model_path
+        )
+        self.model_loader = model_loader
+        self.fingerprint_calculator = fingerprint_calculator
+        self.reaction_encoding_names = reaction_encoding_names
+        self.scalability_encoding = scalability_encoding
+        self.logger = logger
+
+    def reaction_type(self, mol1: str, mol2: str) -> tuple[str, int]:
```
Comment (riya-singh28) [inline deepretro/utils/parse_metrics.py:111]: Add a working example

File: deepretro/utils/parse_metrics.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        """
+        try:
+            return self._predict_reaction_type(mol1, mol2)
+        except FileNotFoundError:
+            return UNKNOWN_REACTION
+        except Exception as exc:
+            self._log_exception("get_reaction_type", exc)
+            return UNKNOWN_REACTION
+
+    def scalability_index(self, mol1: str, mol2: str) -> str:
```
Comment (riya-singh28) [inline deepretro/utils/parse_metrics.py:152]: Add a working example

File: docs/source/package/deepretro.utils.parse.rst [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+- ``format_output`` parses a route tree and returns viewer-ready output.
+
+Example
+-------
+
+.. code-block:: python
+
+   from deepretro.utils.parse import RetrosynthesisRouteParser
+
+   parser = RetrosynthesisRouteParser(
```
Comment (riya-singh28) [inline docs/source/package/deepretro.utils.parse.rst:37]: Please add a better example. Use bigger molecules if required.

File: docs/source/package/deepretro.utils.parse_metrics.rst [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+   calculator = ReactionMetricCalculator(
+       model_path="unused-in-test.joblib",
+       model_loader=lambda path: FakeClassifier(),
+       fingerprint_calculator=lambda smiles: [1, 0],
+       reaction_encoding_names={0: "Example reaction"},
+       scalability_encoding={0: "medium"},
+   )
+
+   assert calculator.reaction_type("CC", "CCO") == ("Example reaction", 0)
```
Comment (riya-singh28) [inline docs/source/package/deepretro.utils.parse_metrics.rst]: Use a better example

**Review by riya-singh28 — CHANGES_REQUESTED** [review]
@Darkxzie I have left a few comments after initial review.

File: deepretro/utils/parse.py (hunk 7 above)
Comment (Darkxzie) [reply deepretro/utils/parse.py:310]: no naming backend was wired in parse.py so it currently empty. if required i can make  changes.

File: docs/source/package/deepretro.utils.parse.rst (hunk 15 above)
Comment (Darkxzie) [reply docs/source/package/deepretro.utils.parse.rst:37]: added.

**Comment by shreyasvinaya** [comment]
closing as content merged as a part of #204
